File: src/main/java/github/lms/lemuel/notification/adapter/in/web/NotificationController.java
package github.lms.lemuel.notification.adapter.in.web;

import github.lms.lemuel.notification.adapter.in.web.response.NotificationResponse;
import github.lms.lemuel.notification.application.port.in.NotificationUseCase;
import github.lms.lemuel.notification.domain.Notification;
import jakarta.validation.constraints.Positive;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Notification API Controller
 */
@Validated
@RestController
@RequestMapping("/api/notifications")
@RequiredArgsConstructor
public class NotificationController {

    private final NotificationUseCase notificationUseCase;

    @GetMapping("/user/{userId}")
    public ResponseEntity<List<NotificationResponse>> getUserNotifications(
            @PathVariable @Positive(message = "유저 ID는 양수여야 합니다") Long userId) {
        List<NotificationResponse> notifications = notificationUseCase.getUserNotifications(userId)
                .stream()
                .map(NotificationResponse::from)
                .collect(Collectors.toList());
        return ResponseEntity.ok(notifications);
    }

    @GetMapping("/user/{userId}/unread")
    public ResponseEntity<List<NotificationResponse>> getUnreadNotifications(
            @PathVariable @Positive(message = "유저 ID는 양수여야 합니다") Long userId) {
        List<NotificationResponse> notifications = notificationUseCase.getUnreadNotifications(userId)
                .stream()
                .map(NotificationResponse::from)
                .collect(Collectors.toList());
        return ResponseEntity.ok(notifications);
    }

    @GetMapping("/user/{userId}/unread/count")
    public ResponseEntity<Long> getUnreadCount(
            @PathVariable @Positive(message = "유저 ID는 양수여야 합니다") Long userId) {
        long count = notificationUseCase.getUnreadCount(userId);
        return ResponseEntity.ok(count);
    }

    @PatchMapping("/{id}/read")
    public ResponseEntity<NotificationResponse> markAsRead(
            @PathVariable @Positive(message = "알림 ID는 양수여야 합니다") Long id) {
        Notification notification = notificationUseCase.markAsRead(id);
        return ResponseEntity.ok(NotificationResponse.from(notification));
    }

    @PatchMapping("/user/{userId}/read-all")
    public ResponseEntity<Void> markAllAsRead(
            @PathVariable @Positive(message = "유저 ID는 양수여야 합니다") Long userId) {
        notificationUseCase.markAllAsRead(userId);
        return ResponseEntity.ok().build();
    }
}

File: notification-service/src/main/kotlin/github/lms/lemuel/notification/adapter/in/web/NotificationController.kt
package github.lms.lemuel.notification.adapter.`in`.web

import github.lms.lemuel.notification.application.ChannelResult
import github.lms.lemuel.notification.application.DispatchResult
import github.lms.lemuel.notification.application.NotificationDispatcher
import github.lms.lemuel.notification.domain.Notification
import github.lms.lemuel.notification.domain.NotificationType
import kotlinx.coroutines.runBlocking
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.UUID

/** Inbound REST request. */
data class SendNotificationRequest(
    val type: String? = null,
    val recipient: String,
    val subject: String,
    val body: String,
    val eventId: String? = null,
)

/** Response DTOs — flat, JSON-friendly. */
data class ChannelResultDto(
    val channel: String,
    val status: String,
    val attempts: Int,
    val error: String? = null,
)

data class DispatchResponse(
    val deduped: Boolean,
    val allSucceeded: Boolean,
    val results: List<ChannelResultDto>,
)

@RestController
@RequestMapping("/notifications")
class NotificationController(
    private val dispatcher: NotificationDispatcher,
) {

    @PostMapping("/send")
    fun send(@RequestBody req: SendNotificationRequest): ResponseEntity<DispatchResponse> {
        val notification = Notification(
            type = parseType(req.type),
            recipient = req.recipient,
            subject = req.subject,
            body = req.body,
            eventId = req.eventId,
        )
        // MVC controller bridges into the coroutine world with runBlocking.
        val result = runBlocking { dispatcher.dispatch(notification) }
        return ResponseEntity.ok(result.toResponse())
    }

    @GetMapping("/demo")
    fun demo(): DispatchResponse {
        val sample = Notification(
            type = NotificationType.SETTLEMENT_CONFIRMED,
            recipient = "ops@lemuel.co.kr",
            subject = "데모 정산 확정",
            body = "notification-service 데모 알림 — 모든 활성 채널로 팬아웃합니다.",
            eventId = "demo-${UUID.randomUUID()}",
        )
        return runBlocking { dispatcher.dispatch(sample) }.toResponse()
    }

    private fun parseType(raw: String?): NotificationType =
        raw?.let {
            runCatching { NotificationType.valueOf(it.uppercase()) }.getOrNull()
        } ?: NotificationType.GENERIC

    private fun DispatchResult.toResponse() = DispatchResponse(
        deduped = deduped,
        allSucceeded = allSucceeded,
        results = results.map { r ->
            when (r) {
                is ChannelResult.Success -> ChannelResultDto(r.channel, "SUCCESS", r.attempts)
                is ChannelResult.Failure -> ChannelResultDto(r.channel, "FAILURE", r.attempts, r.error)
            }
        },
    )
}
